README.md
# Enhance!

<img src="art/logo.png" />

Command line utility to process Android source files distributed via SDK manager and add API version information as javadoc tags:

```java
/**
 * Called by the system when the activity changes from fullscreen mode to multi-window mode and
 * visa-versa.
 *
 * @see android.R.attr#resizeableActivity
 * @param isInMultiWindowMode True if the activity is in multi-window mode.
 * @deprecated Use {@link #onMultiWindowModeChanged(boolean, Configuration)} instead.
 * @since 7.0 Nougat (24)
 * @deprecated 8.0 Oreo (26)
 */
@Deprecated
public void onMultiWindowModeChanged(boolean isInMultiWindowMode) {
    // Left deliberately empty. There should be no side effects if a direct
    // subclass of Activity does not call super.
}
```

Ironically allows to **actually format** processed code with **AOSP** code style specification (and as [google-java-format](https://github.com/google/google-java-format) is used - `GOOGLE` style is also supported).

---

Pick the `jar` file from the latest [release](https://github.com/noties/Enhance/releases/latest/).

There are few configuration options:
* `sdk`: (required) Android SDK version (for example 25)
* `format`: Allows to format processed Java source files. Available options are: `aosp` and `google`. Everything else (including empty argument) won't format processed code
* `sp`: path to Android SDK
* `diff`: just generate statistics info/diff for specified SDK version

```
usage: Enhance
 -diff                  Emit diff
 -format <arg>          Format sources. Accepts (aosp|google). Everything
                        else would keep original formatting
 -h,--help              Prints help
 -sdk <arg>             Specify which SDK version to process.
 -sp,--sdk-path <arg>   Path to Android SDK. If not specified
                        'ANDROID_HOME' system variable will be used
```

Please note that you Android SDK folder must already contain sources for specified `sdk` version.

So, usage would be like that:

```bash
# just add api information to source code, no formatting
java -jar enhance.jar -sdk 26

# also format with AOSP
java -jar enhance.jar -sdk 26 -format aosp

# or GOOGLE
java -jar enhance.jar -sdk 26 -format google

# or with custom SDK path
java -jar enhance.jar -sdk 26 -sp "/Users/not_me/android/sdk"
```

If you would like to restore unmodified copy of source code you can find it: `{your-home-directory}/.enhance-backup/android-{sdk}`

## Formatting on JDK 17
Formatting is done with the [google-java-format](https://github.com/google/google-java-format) library
which requires access to the internals of the JDK. This is why on JDK-17 in order to format
the sources additional commandline arguments are required:

```bash
java \
  --add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED \
  --add-exports=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED \
  --add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED \
  --add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED \
  --add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED \
  --add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED \
  -jar enhance-34-all.jar -sdk 34 -format google
```

## Thanks

Big kudos to the maintainers of amazing [javaparser](https://github.com/javaparser/javaparser)!

## License

```
  Copyright 2018 Dimitry Ivanov (legal@noties.io)

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
```

CHANGELOG.md
# CHANGELOG

# 34
Applied new version-name strategy - now version equals latest supported Android SDK platform
### Added
* Upside Down Cake (34)

### Changed
* Updated code structure, moved to `io.noties` package (no change for clients)


# 1.3.0
### Added
* Android S_V2 (32)
* Tiramisu (33)


# 1.2.0
### Added
* Android 12, SDK 31 API version

# 1.1.1
### Fixed
* Error message when SDK path is missing ([#2])<br>Thanks [@drakeet]

[#2]: https://github.com/noties/Enhance/pull/2
[@drakeet]: https://github.com/drakeet

# 1.1.0
### Added
* Android 11 (R) api version
* Ability to generate just the diff with the `diff` command line option

### Fixed
* signature generation in `ApiInfoStoreImpl` 

src/main/java/io/noties/enhance/Api.java
package io.noties.enhance;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.EnumMap;
import java.util.List;

public enum Api {
    SDK_1(1, "1.0", "(initial)"),
    SDK_2(2, "1.1", "(initial)"),
    SDK_3(3, "1.5", "Cupcake"),
    SDK_4(4, "1.6", "Donut"),
    SDK_5(5, "2.0", "Eclair"),
    SDK_6(6, "2.0.1", "Eclair"),
    SDK_7(7, "2.1", "Eclair"),
    SDK_8(8, "2.2", "Froyo"),
    SDK_9(9, "2.3", "Gingerbread"),
    SDK_10(10, "2.3.3", "Gingerbread"),
    SDK_11(11, "3.0", "Honeycomb"),
    SDK_12(12, "3.1", "Honeycomb"),
    SDK_13(13, "3.2", "Honeycomb"),
    SDK_14(14, "4.0", "Ice Scream Sandwich"),
    SDK_15(15, "4.0.3", "Ice Scream Sandwich"),
    SDK_16(16, "4.1", "Jelly Bean"),
    SDK_17(17, "4.2", "Jelly Bean"),
    SDK_18(18, "4.3", "Jelly Bean"),
    SDK_19(19, "4.4", "Kitkat"),
    SDK_20(20, "4.4W", "Kitkat"),
    SDK_21(21, "5.0", "Lollipop"),
    SDK_22(22, "5.1", "Lollipop"),
    SDK_23(23, "6.0", "Marshmallow"),
    SDK_24(24, "7.0", "Nougat"),
    SDK_25(25, "7.1", "Nougat"),
    SDK_26(26, "8.0", "Oreo"),
    SDK_27(27, "8.1", "Oreo"),
    SDK_28(28, "9.0", "Pie"),
    SDK_29(29, "10", "Android Q"),
    SDK_30(30, "11", "Android R"),
    SDK_31(31, "12", "Android S"),
    SDK_32(32, "12", "Android S_V2"),
    SDK_33(33, "13", "Tiramisu"),
    SDK_34(34, "14", "Upside Down Cake")
    ;

    public final int sdkInt;
    public final String versionName;
    public final String codeName;

    Api(int sdkInt, @Nonnull String versionName, @Nonnull String codeName) {
        this.sdkInt = sdkInt;
        this.versionName = versionName;
        this.codeName = codeName;
    }

    @Override
    public String toString() {
        return "Api.SDK{" +
                "sdkInt=" + sdkInt +
                ", versionName='" + versionName + '\'' +
                ", codeName='" + codeName + '\'' +
                '}';
    }

    private static final List<Api> VALUES = List.of(values());

    @Nonnull
    public static Api latest() {
        return VALUES.get(VALUES.size() - 1);
    }

    @Nullable
    public static Api of(int sdkInt) {
        // NB! we assume that it starts at 1
        // so, SDK_1 is at 0, SDK_2 at 1, etc
        final int ordinal = sdkInt - 1;
        if (ordinal < 0 || ordinal >= VALUES.size()) {
            return null;
        }
        return VALUES.get(ordinal);
    }
}
